### PR TITLE
Add support for Java

### PR DESCRIPTION
**Pull request by pcrock-thmdo**
* adds new image with `jre-21` tag (latest LTS Java release)
* adds new `python-3.12` tag
* fixes a minor incorrect `apt` configuration we had (were using an ubuntu package repo for a debian OS)

unfortunately, the eclipse temurin base container doesn't support debian, just ubuntu. so to support our existing debian bookworm-based python image alongside the ubuntu-based java image, we need to make the `apt` source a docker build arg.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -64,3 +65,26 @@ jobs:
...
+
+  build-java:
+    runs-on: ubuntu-22.04
+    needs: lint
+    strategy:
+      matrix:
+        version:
+          - "21"
+    steps:
+      - uses: actions/checkout@v3
```
Comment (zahidkizmaz) [inline .github/workflows/ci.yml:77]: To match with the rest of the file in main, small update here:
```suggestion
      - uses: actions/checkout@v4
```

**Review by zahidkizmaz — COMMENTED** [review]
Code looks good to me thank you for nice comments. They helped me understand what is going on. For testing I only build the image and made sure it succeeds. Are there any other tests should I do?

**Comment by pcrock-thmdo** [comment]
@zahidkizmaz 

> Are there any other tests should I do?

i am testing with another project; we can consider the java container "alpha" for the moment.

if the build succeeds for the other already-existing containers, that's a good enough test.

thanks a lot :raised_hands:

**Review by zahidkizmaz — APPROVED** [review]

**Review by syphar — APPROVED** [review]

**Comment by syphar** [comment]
ah damn, I was too quick merging, and used the default squashing, while in this example, separate commits would be better...

**Comment by pcrock-thmdo** [comment]
@syphar my fault, shouldn't have bundled 3 things into 1 PR. i still have the commits on my laptop, could do some :cowboy_hat_face: git history rewriting.... :grimacing:

**Comment by syphar** [comment]
feel free, fine with me

**Pull request merged**